src/tesseract.pybind.h

```diff
@@ -83,28 +84,116 @@ void add_tesseract_module(py::module& root) {
+                detections.push_back(i);
+              }
+            }
             self.decode(detections);
+            // Note: `std::vector<bool>` is a special C++ template that does not
+            // provide a contiguous memory block, which is required by `pybind11`
+            // for direct NumPy array creation. Therefose, I use `std::vector<char>`
+            // instead to ensure compatibility with `py::array`.
+            std::vector<char> result(self.num_observables, 0);
             for (size_t ei : self.predicted_errors_buffer) {
               for (int obs_index : self.errors[ei].symptom.observables) {
                 result[obs_index] = result[obs_index] ^ true;
               }
             }
+            return py::array(py::dtype::of<bool>(), result.size(), result.data());
+          },
+          py::arg("syndrome"),
+          py::call_guard<py::scoped_ostream_redirect, py::scoped_estream_redirect>(),
+          R"pbdoc(
+        Decodes a single shot.
+
+        Parameters
+        ----------
+        syndrome : np.ndarray
+            A 1D NumPy array of booleans representing the detection events for a single shot.
+            The length of the array should match the number of detectors in the DEM.
+
+        Returns
+        -------
+        np.ndarray
+            A 1D NumPy array of booleans indicating which observables are flipped.
+            The length of the array matches the number of observables.
+    )pbdoc")
+      .def(
+          "decode_batch",
+          [](TesseractDecoder& self, const py::array_t<bool>& syndromes) {
+            // Ensure the input is in correct 2D array.
+            if (syndromes.ndim() != 2) {
+              throw std::runtime_error("Input syndromes must be a 2D NumPy array.");
+            }
+
+            // Extract dimensions and syndromes.
+            auto syndromes_unchecked = syndromes.unchecked<2>();
+            size_t num_shots = syndromes_unchecked.shape(0);
+            size_t num_detectors = syndromes_unchecked.shape(1);
+
+            // Allocate memory for the function output.
+            py::array_t<bool> result({num_shots, self.num_observables});
+            auto result_unchecked = result.mutable_unchecked<2>();
+
+            // Iterate over each shot and decode it.
+            for (size_t i = 0; i < num_shots; ++i) {
+              std::vector<uint64_t> detections;
+              for (size_t j = 0; j < num_detectors; ++j) {
+                if (syndromes_unchecked(i, j)) {
+                  detections.push_back(j);
+                }
+              }
+              self.decode(detections);
+
+              // Collect results for the current shot being decoded.
+              std::vector<char> shot_result(self.num_observables, 0);
+              for (size_t ei : self.predicted_errors_buffer) {
+                for (int obs_index : self.errors[ei].symptom.observables) {
+                  shot_result[obs_index] ^= true;
```

Comment:
It might be simpler to XOR straight into the numpy array, for example how about something like the following:

```suggestion
                  result_unchecked(i, obs_index) ^= true;
```